---
title: "ProteomicsHW"
author: "Valeria"
date: "2024-04-20"
output: 
  html_document:
    keep_md: yes
---

```{r}
library(readxl)
library(limma)
library(ape)
library(dendextend)
library(RColorBrewer)
library(pvclust)
library(gplots)
library(NMF)
library(vegan)
library(Biobase)
library(DT)
library(ggplot2)
library(impute)
library(ggrepel)
```

## Download data

```{r}
data <- read.csv('peaks_data.csv')
```

## EDA

```{r}
summary(data)
```

Remove genes with half and more missing values and duplicates

```{r}
total_cols <- ncol(data[, !names(data) %in% c("Accession", "Gene_id")])

# Calculate threshold for more than half NaNs
half_threshold <- total_cols / 2

# Identify rows with more than half NaNs
nan_row_counts <- rowSums(is.na(data[, !names(data) %in% c("Accession", "Gene_id")]))

# Get names of rows where NaN count exceeds half the total number of columns
nan_rows <- rownames(data)[nan_row_counts > half_threshold]

data <- data[!rownames(data) %in% nan_rows, ]
data <- data[which(!duplicated(data$Gene_id)), ]

```

## NA values imputation using KNN

```{r}
t_data <- t(as.matrix(data[, !names(data) %in% c("Accession", "Gene_id")]))[-1, ]

knn_data <- impute.knn(t_data, k = 5)
data_matr <- knn_data$data
data_matr <- as.data.frame(data_matr)
data_matr <- t(as.matrix(data_matr))
head(data_matr, 3)

```

```{r}
experiment <- as.factor(sapply(strsplit(colnames(data_matr), "_"), `[`, 1))
pal <- brewer.pal(n = 9, name = "Set1")
cols <- pal[experiment]
```

```{r}
boxplot(data_matr, outline = FALSE, main = "Input data", col = cols)
legend("topright", levels(experiment), fill = pal, bty = "n", xpd = T)

```

##log and norm

```{r}
data_log <- log(data_matr + 1)
data_norm_log <- normalizeQuantiles(as.matrix(data_log))
```

##Evaluate the presence of a batch-effect

```{r}
data_pca <- t(data_norm_log)
terapod_pca <- rda(data_pca, scale = TRUE)

rownames(data_pca) <- experiment

df_scores <- data.frame(data_pca,
                        scores(terapod_pca, display = "sites", choices = c(1, 2, 3), scaling = "sites"))

p_scores <- ggplot(df_scores, aes(x = PC1, y = PC2)) + 
  geom_point(aes(color = rownames(data_pca)), alpha = 0.5) +
  coord_equal(xlim = c(-3, 3), ylim = c(-3, 3))+ 
  ggtitle(label = "PCA") + 
  theme_bw()
p_scores
```

```{r}
maplot <- function(X1, X2, pch = 21, main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", lpars = list(col = "blue", lwd = 2), ...) {
  X <- (rowMeans(X2) + rowMeans(X1)) / 2
  Y <- rowMeans(X2) - rowMeans(X1)
  scatter.smooth(x = X, y = Y, main = main, pch = pch, xlab = xlab, ylab = ylab, lpars = lpars, ...)
  abline(h = c(-1, 0, 1), lty = c(2, 1, 2))
}
maplot(data_norm_log[, 1:14], data_norm_log[, 14:32])
```

## Differential expression

```{r}
expr_data <- as.matrix(data_norm_log)

pheno_data <- data.frame(experiment)
rownames(pheno_data) <- colnames(data_matr)
pheno_metadata <- data.frame(
    labelDescription = c("Experimental condition"),
    row.names = c("Condition")
)
pheno_data <- new("AnnotatedDataFrame",
    data = pheno_data,
    varMetadata = pheno_metadata
)

feature_data <- data.frame(Prot = data$Gene_id)
rownames(feature_data) <- rownames(expr_data)
feature_metadata <- data.frame(
    labelDescription = c("Protain name"),
    row.names = c("Protain")
)
f_data <- new("AnnotatedDataFrame",
    data = feature_data,
    varMetadata = feature_metadata
)

exp_set <-
    ExpressionSet(
        assayData = expr_data,
        phenoData = pheno_data,
        featureData = f_data
    )

```

```{r}
X <- model.matrix(~experiment, pData(exp_set))
fit <- lmFit(exp_set, design = X, method = "robust", maxit = 1000)
efit <- eBayes(fit)
```

```{r}
MA_limma <- function(efit, coef, n = 10, signif = TRUE, fdr = 0.05, lfc = 0, text = TRUE, cex.text = 0.8, col.text = "grey20", main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", pch = 19, pch.signif = 21, col = "darkgreen", alpha = 0.3, cex = 0.3, ...) {
  R <- efit$coefficients[, coef]
  I <- efit$Amean
  col_btransp <- adjustcolor(col, alpha.f = alpha)
  plot(I, R, cex = cex, main = main, pch = pch, xlab = xlab, ylab = ylab, col = col_btransp, ...)
  abline(h = 0)
  if (signif) {
    sign <- p.adjust(efit$p.value[, coef], method = "BH") <= fdr
    large <- abs(efit$coefficients[, coef]) >= lfc
    points(I[sign & large], R[sign & large], cex = cex * 2, col = "orange2", pch = pch.signif)
  }
  if (text) {
    ord <- order(efit$lods[, coef], decreasing = TRUE)
    top_n <- ord[1:n]
    text(I[top_n], R[top_n], labels = efit$genes[top_n, ], pos = 4, cex = cex.text, col = col.text)
  }
}

MA_limma(efit, coef = 2, n = 30)
```

```{r}
my_list <- topTable(efit, coef = 2, n = 100)
dif_exp_set <- exp_set[fData(exp_set)$Prot %in% my_list$Prot, ]
```

```{r}
dat <- as.matrix(exprs(dif_exp_set))
pal_blue_red <- colorpanel(75, low = "steelblue", mid = "black", high = "red")
heatmap.2(dat, col = pal_blue_red, scale = "row", key = TRUE, symkey = FALSE, density.info = "none", trace = "none", cexRow = 0.9, cexCol = 1, margins = c(4, 3), keysize = 0.8, key.par = list(mar = c(3, 0.1, 3, 0.1)))
```

```{r}
topTable(efit, coef = 2)
numGenes <- nrow(exprs(exp_set))
full_list <- topTable(efit, number = numGenes)
full_list <- full_list[full_list$adj.P.Val <= 0.05, ]
write.csv(full_list, "DE_full.csv")
```

## Volcano plot

```{r}
volcano_list <- full_list

volcano1 <- ggplot(data = volcano_list, aes(x = logFC, y = -log10(adj.P.Val))) + geom_point()

volcano2 <- volcano1 + geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")

volcano2
```

```{r message=FALSE, warning=FALSE}
volcano_list$diffexpressed[volcano_list$logFC > 1 & volcano_list$adj.P.Val < 0.05] <- "UP"
```

if log2Foldchange \< -1 and pvalue \< 0.05, set as "DOWN" (можно -0.06)

```{r }
volcano_list$diffexpressed[volcano_list$logFC < -1 & volcano_list$adj.P.Val< 0.05] <- "DOWN"

volcanodif1 <- ggplot(data = volcano_list, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed)) +
  geom_point() + theme_minimal() +
  geom_text_repel(data = head(volcano_list, 20), aes(label = head(volcano_list, 20)$Prot))
volcanodif1
```

## Ordination by DE proteins

```{r fig.height=7, fig.width=15, message=FALSE, warning=FALSE}
diff_prot <- rownames(full_list)
diff_expressed <- as.data.frame(data_norm_log)[diff_prot,]
t_diff_expressed <- t(diff_expressed)
#rownames(t_diff_expressed) <-  as.factor(gsub("_[^_]*", replacement = "", rownames(t_diff_expressed)))
#rownames(t_diff_expressed) <- make.unique(as.character(pheno$yeast_experiment))

d <- dist(x = t_diff_expressed, method = "canberra")

mouse_hc_avg <- hclust(d, method = "average")
mouse_ph_avg <- as.phylo(mouse_hc_avg)
mouse_den_avg <- as.dendrogram(mouse_hc_avg)

get_colours <- function(dend, n_chars, palette = "Dark2"){ #nchars = первые нескольок симовлов которые будут использоваться для разделения фактора на группы
labs <- get_leaves_attr(dend, "label")
group <- substr(labs, start = 0, stop = n_chars)
group <- factor(group)
cols <- brewer.pal(length(levels(group)), name = palette)[group]
return(cols)
}

cols <- get_colours(dend = mouse_den_avg, n_chars = 6)
den_avg_c <- color_labels(dend = mouse_den_avg, col = cols)
plot(den_avg_c, horiz = TRUE)
```

## GO enrichment analysis

```{r}
install.packages("gprofiler2")
library(gprofiler2)
```

```{r}
library(clusterProfiler)
library(org.Hs.eg.db)
```

```{r}
sig <- subset(full_list, adj.P.Val < 0.05)
up <- subset(sig, logFC > 3)
down <- subset(sig, logFC < -2)
up_names <- gconvert(up$Prot)
down_names <- gconvert(down$Prot)
```

## Up-regulated proteins

```{r}

multi_gp_up_reg <- gost(list("up-regulated" = up_names), multi_query = FALSE, evcodes =TRUE)
# modify the g:Profiler data frame
gp_mod_up = multi_gp_up_reg$result[, c("query", "source", "term_id","term_name", "p_value", "query_size", "intersection_size", "term_size", "effective_domain_size", "intersection")]
gp_mod_up <- gp_mod_up[order(gp_mod_up$p_value), ]
gp_mod_up_BP <- gp_mod_up[gp_mod_up$source == "GO:BP", ]
gp_mod_up_BP$GeneRatio <- paste0(gp_mod_up_BP$intersection_size,  "/", gp_mod_up_BP$query_size)
gp_mod_up_BP$BgRatio <- paste0(gp_mod_up_BP$term_size, "/", gp_mod_up_BP$effective_domain_size)
names(gp_mod_up_BP) <- c("Cluster", "Category", "ID", "Description", "p.adjust", "query_size", "Count", "term_size", "effective_domain_size", "geneID", "GeneRatio", "BgRatio")
gp_mod_up_BP$geneID <- gsub(",", "/", gp_mod_up_BP$geneID)
row.names(gp_mod_up_BP) <- gp_mod_up_BP$ID
gp_mod_enrich_up_BP <- new("enrichResult", result = gp_mod_up_BP)


enrichplot::dotplot(gp_mod_enrich_up_BP, showCategory = 10) + ggplot2::labs(title = "up-regulated") + ggplot2::scale_color_gradient(low = "lightseagreen", high = "darkorange1")


```

```{r}
# enrichment analysis using gene names
multi_gp_down_reg <- gost(list("down-regulated" = down_names), multi_query = FALSE, evcodes =TRUE)
# modify the g:Profiler data frame
gp_mod_down = multi_gp_down_reg$result[, c("query", "source", "term_id","term_name", "p_value", "query_size", "intersection_size", "term_size", "effective_domain_size", "intersection")]
gp_mod_down <- gp_mod_down[order(gp_mod_down$p_value), ]
# BP
gp_mod_down_BP <- gp_mod_down[gp_mod_down$source == "GO:BP", ]
gp_mod_down_BP$GeneRatio <- paste0(gp_mod_down_BP$intersection_size,  "/", gp_mod_down_BP$query_size)
gp_mod_down_BP$BgRatio <-  paste0(gp_mod_down_BP$term_size, "/", gp_mod_down_BP$effective_domain_size)
names(gp_mod_down_BP) <- c("Cluster", "Category", "ID", "Description", "p.adjust", "query_size", "Count", "term_size", "effective_domain_size", "geneID", "GeneRatio", "BgRatio")
gp_mod_down_BP$geneID <- gsub(",", "/", gp_mod_down_BP$geneID)
gp_mod_enrich_down_BP <- new("enrichResult", result = gp_mod_down_BP)
enrichplot::dotplot(gp_mod_enrich_down_BP, showCategory = 10) + ggplot2::labs(title = "down-regulated") + ggplot2::scale_color_gradient(low = "lightseagreen", high = "darkorange1")
```

## Conclusion

I analyzed the differences with a big LFC threshold (logFC \> 3) and (logFC \> -2). We can see that mostly up-regulated genes are connected with carboxilic and catabolic processes.This can lead to a condition where when the rate of protein degradation exceeds the rate of protein synthesis [1]. The major catabolic pathways downstream of these include the ubiquitinproteasome system (1), the lysosomal system (2), Ca2+- dependent calpains(3) and caspases (4). On the volcano plot we can see particular genes involved in these procceses. For example, UBASH3B is encodes a protein that contains a ubiquitin associated domain at the N-terminus, an SH3 domain, and a C-terminal domain with similarities to the catalytic motif of phosphoglycerate mutase (1). The encoded protein was found to inhibit endocytosis of epidermal growth factor receptor (EGFR) and platelet-derived growth factor receptor [2]. LTN1 functions as an E3 ubiquitin ligase [3]. The pathways that showed a significant enrichment in the CJK group compared to the control group indicate the presence of oxidative stress. On the other hand, we see that in addition to catabolic processes, oxidation processes appear in down regulated genes, which is also an indicator of oxidative stress [4].

1.  <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2877703/>
2.  <https://www.genecards.org/cgi-bin/carddisp.pl?gene=UBASH3B>
3.  <https://www.genecards.org/cgi-bin/carddisp.pl?gene=LTN1>
4.  <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4136416/>
